flybywiresim/a32nx
# feat(atsu): VDL3 simulation

## Summary of Changes

The real datalink uses in VHF-areas VDL2 or VDL3 communication protocols to exchange ACARS/CPDLC data between ground stations and air stations. The current implementation works with fixed timings to simulate transmission delays.

This PR implements a VDL3-simulation (due to better simulation capabilities) and includes a VHF signal quality simulation.
SITA and ARINC are used as simulated datalink providers with one frequency for the world. The current implementation assumes that every international airport has SITA and ARINC stations which can be used for the datalink communication.

The VHF signal quality simulation is based on line-of-sight and distance based signal damping estimations and frequency overlapping to predict the data rate for the communication protocol. Due to a missing HF implementation is VHF never deactivated, but decreases the data rate to 62 bytes per second. The VDL3 simulation takes all ground-stations and air-stations into account and shares the data slots equally between all stations.

Additionally are request-response (i.e. weather or ATIS) better simulated by use three realistic stages:
- send request with VDL3-simulation
- wait some time (300 - 800 ms) to simulate a processing time on the ground station
- receive response with VDL3-simulation

An other change/fix is the handover-handling from one ATC to an other.
The system checks if open messages (that need responses or need to be sent) are buffered in the DCDU.
The performed handover is postponed until all messages are sent to the current ATC.
Thereafter is the handover performed.

The last change is that the CONNECT-button in the EFB is removed.
The system checks if VATSIM or IVAO is used as METAR and ATIS source and the Hoppie logon is valid.

## Screenshots (if necessary)




## References


VDL3: https://en.wikipedia.org/wiki/VHF_Data_Link#ICAO_VDL_Mode_3
LoS-estimation: https://audio.vatsim.net/storage/AFV%20User%20Guide.pdf




## Additional context



Discord username (if different from GitHub):

## Testing instructions

- open http://localhost:19999 and check the log of the CDU
- Test if Hoppie is active if the following conditions are fulfilled (printed in log of CDU or reason why it is not active)
  - Logon code in EFB
  - METAR and ATIS source as IVAO or VATSIM
- Test that all requests (weather/atis) are still working
- Test that freetext is still working
- Test that Connection Notification and Connection Status are still working
- Possible checks of VDL3-simulation:
  - use AI traffic
  - shows every 30 seconds the number of stations, relevant airports and the time between to packets
- Test handover
  - Handover without open messages in the DCDU should be performed within the transmission delays
  - Handover with open messages should be postponed until the DCDU message is answered/sent
    - Handover waits 15 seconds after the last open message of the DCDU and is performed thereafter
- Test different transmission speeds with METAR for four airports
  - One were no relevant airport is found (i.e. KFBW) and one were relevant airports are found (i.e. KSFO)
  - Transmission should be faster when relevant airports are found

Example for the Hoppie activation:
![grafik](https://user-images.githubusercontent.com/64070348/155223840-3f238841-c85c-43c7-853d-7bd59c593315.png)

Example of the VDL3 simulation:
![grafik](https://user-images.githubusercontent.com/64070348/155223959-56715596-1681-488d-8424-611b12b39fcc.png)


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
